packages/loopring_v3/contracts/amm/libamm/AmmJoinProcess.sol

```diff
@@ -0,0 +1,211 @@
+library AmmJoinProcess
+{
+    using AmmPoolToken      for AmmData.State;
+    using AmmStatus         for AmmData.State;
+    using AmmUtil           for uint96;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+    using TransactionReader for ExchangeData.Block;
+
+    function proxcessExchangeDeposit(
+        AmmData.State    storage S,
+        AmmData.Context  memory  ctx,
+        AmmData.Token    memory  token,
+        uint96                   amount
+        )
+        internal
+    {
+        require(amount > 0, "INVALID_DEPOSIT_AMOUNT");
+
+        // Check that the deposit in the block matches the expected deposit
+        DepositTransaction.Deposit memory deposit = ctx._block.readDeposit(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        require(
+            deposit.owner == address(this) &&
+            deposit.accountID == S.accountID &&
+            deposit.tokenID == token.tokenID &&
+            // deposit.amount.isAlmostEqual(amount),
+            deposit.amount == amount,
+            "INVALID_TX_DATA"
+        );
+
+        // Now do this deposit
+        uint ethValue = 0;
+        if (token.addr == address(0)) {
+            ethValue = amount;
+        } else {
+            uint allowance = ERC20(token.addr).allowance(address(this), ctx.exchangeDepositContract);
+            if (allowance < amount) {
+                // Approve the deposit transfer
+                ERC20(token.addr).approve(ctx.exchangeDepositContract, uint(-1));
+            }
+        }
+
+        ctx.exchange.deposit{value: ethValue}(
+            deposit.owner,
+            deposit.owner,
+            token.addr,
+            deposit.amount,
+            new bytes(0)
+        );
+
+        // Total balance in this contract decreases by the amount deposited
+        S.totalLockedBalance[token.addr] = S.totalLockedBalance[token.addr].sub(amount);
+    }
+
+    function processJoin(
+        AmmData.State    storage S,
+        AmmData.Context  memory  ctx,
+        AmmData.PoolJoin memory  join,
+        bytes            memory  signature
+        )
+        internal
+    {
+        S.validatePoolTransaction(
+            join.owner,
+            AmmUtil.hashPoolJoin(ctx.domainSeperator, join),
+            signature
+        );
+
+        // Check if the requirements are fulfilled
+        (bool slippageRequirementMet, uint poolAmountOut, uint96[] memory amounts) = _calculateJoinAmounts(ctx, join);
+
+        if (!slippageRequirementMet) return;
+
+        for (uint i = 0; i < ctx.size; i++) {
+            uint96 amount = amounts[i];
+
+            if (join.fromLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                ctx.numTransactionsConsumed++;
+
+                // We do not check these fields: fromAccountID, to, amount, feeTokenID, fee
```

Comment:
`transfer.fee` should be checked to be something that the user approves, now the operator can just fill in any fee and drain the user's account. So if we want these to be non-zero, the fee will have to be part of the join request that the user approves.

Reply:
I'll enable non-zero fees later. Added two TODOs